home-application/src/main/java/com/softserveinc/ita/homeproject/application/service/QueryApiService.java

```diff
@@ -0,0 +1,111 @@
+package com.softserveinc.ita.homeproject.application.service;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.UriInfo;
+
+import com.softserveinc.ita.homeproject.homedata.entity.BaseEntity;
+import com.softserveinc.ita.homeproject.homeservice.dto.BaseDto;
+import com.softserveinc.ita.homeproject.homeservice.service.QueryableService;
+import org.springframework.data.domain.Page;
+import org.springframework.data.jpa.domain.Specification;
+
+/**
+ * QueryApiService - service that provides Spring Data Page by request query
+ *
+ * @author Oleksii Zinkevych
+ * @param <T> - entity type
+ * @param <D> - DTO type
+ * @see javax.ws.rs.core.UriInfo
+ */
+public interface QueryApiService<T extends BaseEntity, D extends BaseDto> {
+    Set<String> EXCLUDED_PARAMETERS = new HashSet<>(Arrays.stream(DefaultQueryParams.values())
+            .map(DefaultQueryParams::getParameter)
+            .collect(Collectors.toList()));
+
+    /**
+     *
+     * @param uriInfo - object that implements UriInfo interface
+     *                  that provides access to application and request URI information
+     * @return Map of query and path parameters required by EntitySpecificationService
+     * to build Filter specification
+     */
+    static MultivaluedMap<String, String> getFilterMap(UriInfo uriInfo) {
+        MultivaluedMap<String, String> filterMap = new MultivaluedHashMap<>();
+        MultivaluedMap<String, String> queryParameters = uriInfo.getQueryParameters();
+        MultivaluedMap<String, String> pathParameters = uriInfo.getPathParameters();
+        queryParameters.forEach((key, value) -> {
+            if(!(EXCLUDED_PARAMETERS.contains(key))) {
+                filterMap.put(key, value);
+            }
+        });
+        pathParameters.forEach(filterMap::put);
+        return filterMap;
+    }
+
+    /**
+     *
+     * @param uriInfo - object that implements UriInfo interface
+     *                  that provides access to application and request URI information
+     * @return Spring Data Specification of entity
+     */
+    Specification<T> getSpecification(UriInfo uriInfo);
+
+    /**
+     *
+     * @param param   - String name of query parameter
+     * @param uriInfo - object that implements UriInfo interface
+     *                  that provides access to application and request URI information
+     * @return String value of specified query parameter
+     */
+    static String getParameterValue(String param, UriInfo uriInfo) {
```

Comment:
We also have to consider the situation when more than 1 special query param present in the request and validate it as a client error as cannot randomly take the parameter.

`/users?pageNum=2&pageNum=3&pageNum=4` - it should be validated complaining that only one `pageNum` can be present in the request.

Consider the following:

```java
    static String getParameterValue(String param, UriInfo uriInfo) {
        String paramValue = null;
        MultivaluedMap<String, String> queryParameters = uriInfo.getQueryParameters();
        if (queryParameters.containsKey(param)) {
            List<String> paramValues = queryParameters.get(param);
            switch (paramValues.size()) {
                case 0: // Is it possible for this Map?
                    paramValue = null;
                    break;
                case 1:
                    paramValue = paramValues.get(0);
                    break;
                default:
                    throw new IllegalArgumentException(); //TODO client error
            }
        }
        return paramValue;
    }
```